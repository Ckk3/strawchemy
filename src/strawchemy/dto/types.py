"""DTO domain types."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from collections.abc import Callable, Mapping


__all__ = ("DTOConfig", "DTOFieldConfig", "Purpose")


class DTOMissingType:
    """A sentinel type to detect if a parameter is supplied or not when.

    constructing pydantic FieldInfo.
    """


DTO_MISSING = DTOMissingType()


class Purpose(str, Enum):
    """For identifying the purpose of a DTO to the factory.

    The factory will exclude fields marked as private or read-only on the domain model depending
    on the purpose of the DTO.

    Example:
    ```python
    ReadDTO = dto.factory("AuthorReadDTO", Author, purpose=dto.Purpose.READ)
    ```
    """

    READ = "read"
    """To mark a DTO that is to be used to serialize data returned to
    clients."""
    WRITE = "write"
    """To mark a DTO that is to deserialize and validate data provided by
    clients."""
    COMPLETE = "complete"
    """To mark a DTO that is to deserialize and validate data provided by
    clients. Fields marked as TO_COMPLETE must not be null."""


@dataclass
class PurposeConfig:
    """Mark the field as read-only, or private."""

    type_override: Any | None = DTO_MISSING
    validator: Callable[[Any], Any] | None = None
    """Single argument callables that are defined on the DTO as validators for the field."""
    alias: str | None = None
    """Customize name of generated pydantic field."""
    partial: bool | None = None


@dataclass
class DTOFieldConfig:
    """For configuring DTO behavior on SQLAlchemy model fields."""

    purposes: set[Purpose] = field(default_factory=lambda: {Purpose.READ, Purpose.WRITE})
    default_config: PurposeConfig = field(default_factory=PurposeConfig)
    configs: dict[Purpose, PurposeConfig] = field(default_factory=dict)

    def purpose_config(self, dto_config: DTOConfig) -> PurposeConfig:
        return self.configs.get(dto_config.purpose, self.default_config)


@dataclass
class DTOConfig:
    """Control the generated DTO."""

    purpose: Purpose
    """Configure the DTO for "read" or "write" operations."""
    exclude: set[str] = field(default_factory=set)
    """Explicitly exclude fields from the generated DTO."""
    partial: bool | None = None
    """Make all field optional."""
    type_map: Mapping[Any, Any] = field(default_factory=dict)

    aliases: Mapping[str, str] = field(default_factory=dict)

    alias_generator: Callable[[str], str] | None = None

    def __post_init__(self) -> None:
        if self.aliases and self.alias_generator is not None:
            msg = "You must set `aliases` or `alias_generator`, not both"
            raise ValueError(msg)

    def alias(self, name: str) -> str | None:
        if self.aliases:
            return self.aliases.get(name)
        if self.alias_generator is not None:
            return self.alias_generator(name)
        return None

    @property
    def cache_key(self) -> int:
        return hash(
            (self.purpose, self.partial, self.alias_generator, tuple(self.type_map), tuple(self.type_map.values()))
        )
