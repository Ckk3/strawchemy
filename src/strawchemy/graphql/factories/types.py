from __future__ import annotations

import dataclasses
from collections.abc import Generator
from typing import TYPE_CHECKING, Any, TypeVar, override

from sqlalchemy.orm import DeclarativeBase, QueryableAttribute
from strawchemy.dto.backend.strawberry import StrawberrryDTOBackend
from strawchemy.dto.base import (
    DTOBackend,
    DTOBase,
    DTOFactory,
    DTOFieldDefinition,
    Relation,
)
from strawchemy.dto.types import DTO_MISSING, DTOConfig, Purpose
from strawchemy.graph import Node
from strawchemy.graphql import typing as strawchemy_typing
from strawchemy.graphql.constants import AGGREGATIONS_KEY, NODES_KEY
from strawchemy.graphql.dto import (
    AggregateDTO,
    AggregateFieldDefinition,
    DTOKey,
    EnumDTO,
    FunctionFieldDefinition,
    GraphQLFieldDefinition,
)
from strawchemy.graphql.typing import AggregateDTOT, GraphQLDTOT

from .aggregations import AggregationInspector
from .enum import EnumDTOFactory

if TYPE_CHECKING:
    from collections.abc import Generator

    from strawchemy.graph import Node
    from strawchemy.graphql.inspector import GraphQLInspectorProtocol
    from strawchemy.sqlalchemy.typing import DeclarativeT


__all__ = ("AggregateDTOFactory", "DistinctOnFieldsDTOFactory", "RootAggregateTypeDTOFactory", "TypeDTOFactory")

T = TypeVar("T")

_TYPING_NS = vars(strawchemy_typing)


class GraphQLDTOFactory(DTOFactory[DeclarativeBase, QueryableAttribute[Any], GraphQLDTOT]):
    inspector: GraphQLInspectorProtocol

    def type_description(self) -> str:
        return "GraphQL type"

    @override
    def type_hint_namespace(self) -> dict[str, Any]:
        return super().type_hint_namespace() | _TYPING_NS

    @override
    def iter_field_definitions(
        self,
        name: str,
        model: type[T],
        dto_config: DTOConfig,
        base: type[DTOBase[DeclarativeBase]] | None,
        node: Node[Relation[DeclarativeBase, GraphQLDTOT], None],
        raise_if_no_fields: bool = False,
        *,
        field_map: dict[DTOKey, GraphQLFieldDefinition] | None = None,
        **kwargs: Any,
    ) -> Generator[DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]], None, None]:
        field_map = field_map if field_map is not None else {}
        for field in super().iter_field_definitions(name, model, dto_config, base, node, raise_if_no_fields, **kwargs):
            key = DTOKey.from_dto_node(node)
            graphql_field = GraphQLFieldDefinition.from_field(field)
            yield graphql_field
            field_map[key + field.name] = graphql_field

    @override
    def factory(
        self,
        model: type[T],
        dto_config: DTOConfig,
        base: type[Any] | None = None,
        name: str | None = None,
        parent_field_def: DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]] | None = None,
        current_node: Node[Relation[Any, GraphQLDTOT], None] | None = None,
        raise_if_no_fields: bool = False,
        backend_kwargs: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> type[GraphQLDTOT]:
        field_map: dict[DTOKey, GraphQLFieldDefinition] = {}
        dto = super().factory(
            model,
            dto_config,
            base,
            name,
            parent_field_def,
            current_node,
            raise_if_no_fields,
            backend_kwargs,
            field_map=field_map,
            **kwargs,
        )
        if not dto.__strawchemy_field_map__:
            dto.__strawchemy_field_map__ = field_map
        dto.__strawchemy_description__ = self.type_description()
        return dto


class TypeDTOFactory(GraphQLDTOFactory[GraphQLDTOT]):
    def __init__(
        self,
        inspector: GraphQLInspectorProtocol,
        backend: DTOBackend[GraphQLDTOT],
        handle_cycles: bool = True,
        type_map: dict[Any, Any] | None = None,
        aggregation_factory: AggregateDTOFactory[AggregateDTOT] | None = None,
        **kwargs: Any,
    ) -> None:
        super().__init__(inspector, backend, handle_cycles, type_map, **kwargs)
        self._aggregation_factory = aggregation_factory or AggregateDTOFactory(
            inspector, StrawberrryDTOBackend(AggregateDTO)
        )

    def _aggregation_field(
        self, field_def: DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]], dto_config: DTOConfig
    ) -> GraphQLFieldDefinition:
        related_model = self.inspector.relation_model(field_def.model_field)
        aggregate_dto_config = dataclasses.replace(dto_config, annotation_overrides={})
        dto = self._aggregation_factory.factory(
            model=related_model, dto_config=aggregate_dto_config, parent_field_def=field_def
        )
        return AggregateFieldDefinition(
            dto_config=dto_config,
            model=dto.__dto_model__,
            _model_field=field_def.model_field,
            model_field_name=f"{field_def.name}_aggregate",
            type_hint=dto,
            related_dto=dto,
        )

    @override
    def dto_name(
        self, base_name: str, dto_config: DTOConfig, node: Node[Relation[Any, GraphQLDTOT], None] | None = None
    ) -> str:
        return f"{base_name}{'Input' if dto_config.purpose is Purpose.WRITE else ''}Type"

    @override
    def iter_field_definitions(
        self,
        name: str,
        model: type[T],
        dto_config: DTOConfig,
        base: type[DTOBase[DeclarativeBase]] | None,
        node: Node[Relation[DeclarativeBase, GraphQLDTOT], None],
        raise_if_no_fields: bool = False,
        *,
        aggregations: bool = False,
        field_map: dict[DTOKey, GraphQLFieldDefinition] | None = None,
        **kwargs: Any,
    ) -> Generator[DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]], None, None]:
        field_map = field_map if field_map is not None else {}
        for field in super().iter_field_definitions(
            name, model, dto_config, base, node, raise_if_no_fields, field_map=field_map, **kwargs
        ):
            key = DTOKey.from_dto_node(node)
            if field.is_relation and field.uselist and aggregations:
                aggregation_field = self._aggregation_field(field, dto_config)
                field_map[key + aggregation_field.name] = aggregation_field
                yield aggregation_field
            yield field

    @override
    def factory(
        self,
        model: type[T],
        dto_config: DTOConfig,
        base: type[Any] | None = None,
        name: str | None = None,
        parent_field_def: DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]] | None = None,
        current_node: Node[Relation[Any, GraphQLDTOT], None] | None = None,
        raise_if_no_fields: bool = False,
        backend_kwargs: dict[str, Any] | None = None,
        *,
        aggregations: bool = True,
        **kwargs: Any,
    ) -> type[GraphQLDTOT]:
        return super().factory(
            model,
            dto_config,
            base,
            name,
            parent_field_def,
            current_node,
            raise_if_no_fields,
            backend_kwargs,
            aggregations=aggregations if dto_config.purpose is Purpose.READ else False,
            **kwargs,
        )


class RootAggregateTypeDTOFactory(TypeDTOFactory[GraphQLDTOT]):
    def __init__(
        self,
        inspector: GraphQLInspectorProtocol,
        backend: DTOBackend[GraphQLDTOT],
        handle_cycles: bool = True,
        type_map: dict[Any, Any] | None = None,
        type_factory: TypeDTOFactory[GraphQLDTOT] | None = None,
        aggregation_factory: AggregateDTOFactory[AggregateDTOT] | None = None,
        **kwargs: Any,
    ) -> None:
        super().__init__(inspector, backend, handle_cycles, type_map, **kwargs)
        self._type_factory = type_factory or TypeDTOFactory(inspector, backend)
        self._aggregation_factory = aggregation_factory or AggregateDTOFactory(
            inspector, StrawberrryDTOBackend(AggregateDTO)
        )

    @override
    def dto_name(
        self, base_name: str, dto_config: DTOConfig, node: Node[Relation[Any, GraphQLDTOT], None] | None = None
    ) -> str:
        return f"{base_name}Root"

    @override
    def iter_field_definitions(
        self,
        name: str,
        model: type[DeclarativeT],
        dto_config: DTOConfig,
        base: type[DTOBase[DeclarativeBase]] | None,
        node: Node[Relation[DeclarativeBase, GraphQLDTOT], None],
        raise_if_no_fields: bool = False,
        aggregations: bool = False,
        field_map: dict[DTOKey, GraphQLFieldDefinition] | None = None,
        **kwargs: Any,
    ) -> Generator[DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]], None, None]:
        if not node.is_root:
            yield from ()
        key = DTOKey.from_dto_node(node)
        field_map = field_map if field_map is not None else {}
        nodes_dto = self._type_factory.factory(model, dto_config=dto_config, aggregations=aggregations)
        nodes = GraphQLFieldDefinition(
            dto_config=dto_config,
            model=model,
            model_field_name=NODES_KEY,
            type_hint=list[nodes_dto],
            is_relation=False,
        )
        aggregations_field = GraphQLFieldDefinition(
            dto_config=dto_config,
            model=model,
            model_field_name=AGGREGATIONS_KEY,
            type_hint=self._aggregation_factory.factory(model, dto_config=dto_config),
            is_relation=False,
            is_aggregate=True,
        )
        field_map[key + nodes.name] = nodes
        field_map[key + aggregations_field.name] = aggregations_field
        yield from iter((nodes, aggregations_field))

    @override
    def factory(
        self,
        model: type[T],
        dto_config: DTOConfig,
        base: type[Any] | None = None,
        name: str | None = None,
        parent_field_def: DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]] | None = None,
        current_node: Node[Relation[Any, GraphQLDTOT], None] | None = None,
        raise_if_no_fields: bool = False,
        backend_kwargs: dict[str, Any] | None = None,
        *,
        aggregations: bool = True,
        **kwargs: Any,
    ) -> type[GraphQLDTOT]:
        dto = super().factory(
            model,
            dto_config,
            base,
            name,
            parent_field_def,
            current_node,
            raise_if_no_fields,
            backend_kwargs,
            aggregations=aggregations,
            **kwargs,
        )
        dto.__strawchemy_is_root_aggregation_type__ = True
        return dto


class AggregateDTOFactory(GraphQLDTOFactory[AggregateDTOT]):
    def __init__(
        self,
        inspector: GraphQLInspectorProtocol,
        backend: DTOBackend[AggregateDTOT],
        handle_cycles: bool = True,
        type_map: dict[Any, Any] | None = None,
        aggregation_builder: AggregationInspector | None = None,
        **kwargs: Any,
    ) -> None:
        super().__init__(inspector, backend, handle_cycles, type_map, **kwargs)
        self._aggregation_builder = aggregation_builder or AggregationInspector(inspector)

    @override
    def type_description(self) -> str:
        return "Aggregation fields"

    @override
    def dto_name(
        self, base_name: str, dto_config: DTOConfig, node: Node[Relation[Any, AggregateDTOT], None] | None = None
    ) -> str:
        return f"{base_name}Aggregate"

    @override
    def _factory(
        self,
        name: str,
        model: type[DeclarativeT],
        dto_config: DTOConfig,
        node: Node[Relation[Any, AggregateDTOT], None],
        base: type[Any] | None = None,
        parent_field_def: DTOFieldDefinition[DeclarativeBase, QueryableAttribute[Any]] | None = None,
        raise_if_no_fields: bool = False,
        backend_kwargs: dict[str, Any] | None = None,
        field_map: dict[DTOKey, GraphQLFieldDefinition] | None = None,
        **kwargs: Any,
    ) -> type[AggregateDTOT]:
        field_map = field_map if field_map is not None else {}
        model_field = parent_field_def.model_field if parent_field_def else None
        as_partial_config = dataclasses.replace(dto_config, partial=True)
        field_definitions: list[FunctionFieldDefinition] = [
            FunctionFieldDefinition(
                dto_config=dto_config,
                model=model,
                _model_field=model_field if model_field is not None else DTO_MISSING,
                model_field_name=aggregation.function,
                type_hint=aggregation.output_type,
                _function=aggregation,
                default=aggregation.default,
            )
            for aggregation in self._aggregation_builder.output_functions(model, as_partial_config)
        ]

        root_key = DTOKey.from_dto_node(node)
        field_map.update({root_key + field.model_field_name: field for field in field_definitions})
        return self.backend.build(name, model, field_definitions, **(backend_kwargs or {}))


class DistinctOnFieldsDTOFactory(EnumDTOFactory):
    @override
    def dto_name(
        self, base_name: str, dto_config: DTOConfig, node: Node[Relation[Any, EnumDTO], None] | None = None
    ) -> str:
        return f"{base_name}DistinctOnFields"
