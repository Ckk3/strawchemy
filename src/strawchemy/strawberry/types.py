from __future__ import annotations

from enum import Enum
from typing import TYPE_CHECKING, Any, TypeVar

import strawberry
from strawberry import UNSET
from strawchemy.dto.base import MappedDTO
from strawchemy.graphql.mutation import (
    RequiredToManyUpdateInputMixin,
    RequiredToOneInputMixin,
    ToManyCreateInputMixin,
    ToManyUpdateInputMixin,
    ToOneInputMixin,
)
from strawchemy.utils import snake_to_lower_camel_case

if TYPE_CHECKING:
    from pydantic import ValidationError
    from pydantic_core import ErrorDetails


T = TypeVar("T", bound=MappedDTO[Any])
RelationInputT = TypeVar("RelationInputT", bound=MappedDTO[Any])
_TO_ONE_DESCRIPTION = "Add a new or existing object"
_TO_MANY_DESCRIPTION = "Add new or existing objects"
_TO_MANY_UPDATE_DESCRIPTION = "Add new objects or update existing ones"


class ErrorId(Enum):
    ERROR = "ERROR"
    VALIDATION_ERROR = "VALIDATION_ERROR"
    LOCALIZED_VALIDATION_ERROR = "LOCALIZED_VALIDATION_ERROR"


@strawberry.input(description=_TO_ONE_DESCRIPTION)
class ToOneInput(ToOneInputMixin[T, RelationInputT]):
    set: T | None = UNSET
    create: RelationInputT | None = UNSET


@strawberry.input(description=_TO_ONE_DESCRIPTION)
class RequiredToOneInput(RequiredToOneInputMixin[T, RelationInputT]):
    set: T | None = UNSET
    create: RelationInputT | None = UNSET


@strawberry.input(description=_TO_MANY_DESCRIPTION)
class ToManyCreateInput(ToManyCreateInputMixin[T, RelationInputT]):
    set: list[T] | None = UNSET
    add: list[T] | None = UNSET
    create: list[RelationInputT] | None = UNSET


@strawberry.input(description=_TO_MANY_UPDATE_DESCRIPTION)
class ToManyUpdateInput(ToManyUpdateInputMixin[T, RelationInputT]):
    set: list[T] | None = UNSET
    add: list[T] | None = UNSET
    remove: list[T] | None = UNSET
    create: list[RelationInputT] | None = UNSET


@strawberry.input(description=_TO_MANY_UPDATE_DESCRIPTION)
class RequiredToManyUpdateInput(RequiredToManyUpdateInputMixin[T, RelationInputT]):
    set: list[T] | None = UNSET
    add: list[T] | None = UNSET
    create: list[RelationInputT] | None = UNSET


@strawberry.interface(description="Base interface for expected errors", name="ErrorType")
class ErrorType:
    """Base class for GraphQL errors."""

    id: str = ErrorId.ERROR.value


@strawberry.type(description="Indicate validation error type and location.", name="LocalizedErrorType")
class LocalizedErrorType(ErrorType):
    """Match inner shape of pydantic ValidationError."""

    id = ErrorId.LOCALIZED_VALIDATION_ERROR.value
    loc: list[str] = strawberry.field(default_factory=list)
    message: str
    type: str


@strawberry.type(description="Input is malformed or invalid.", name="ValidationErrorType")
class ValidationErrorType(ErrorType):
    """Input is malformed or invalid."""

    id = ErrorId.VALIDATION_ERROR.value
    errors: list[LocalizedErrorType]

    @classmethod
    def _to_localized_error(cls, errors: ErrorDetails, to_camel: bool) -> LocalizedErrorType:
        return LocalizedErrorType(
            loc=[snake_to_lower_camel_case(str(loc)) if to_camel else str(loc) for loc in errors["loc"]],
            message=errors["msg"],
            type=errors["type"],
        )

    @classmethod
    def from_pydantic_error(cls, exc: ValidationError, to_camel: bool = True) -> ValidationErrorType:
        return ValidationErrorType(errors=[cls._to_localized_error(err, to_camel) for err in exc.errors()])
