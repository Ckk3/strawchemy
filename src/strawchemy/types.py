from __future__ import annotations

from dataclasses import dataclass

__all__ = ("DefaultOffsetPagination",)


@dataclass(eq=True, frozen=True)
class DefaultOffsetPagination:
    limit: int | None = None
    offset: int = 0
