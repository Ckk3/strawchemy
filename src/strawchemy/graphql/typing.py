from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal, TypeAlias, TypeVar

if TYPE_CHECKING:
    from collections.abc import Callable

    from .dto import (
        AggregateDTO,
        FilterFunctionInfo,
        GraphQLFilterDTO,
        MappedDataclassGraphQLDTO,
        MappedPydanticGraphQLDTO,
        OrderByDTO,
        OutputFunctionInfo,
        UnmappedDataclassGraphQLDTO,
    )


T = TypeVar("T")
QueryObject = TypeVar("QueryObject", bound=Any)
GraphQLFilterDTOT = TypeVar("GraphQLFilterDTOT", bound="GraphQLFilterDTO")
AggregateDTOT = TypeVar("AggregateDTOT", bound="AggregateDTO")
GraphQLDTOT = TypeVar("GraphQLDTOT", bound="GraphQLDTO[Any]")
OrderByDTOT = TypeVar("OrderByDTOT", bound="OrderByDTO")

AggregationFunction = Literal["min", "max", "sum", "avg", "count", "stddev_samp", "stddev_pop", "var_samp", "var_pop"]
AggregationType = Literal[
    "sum", "numeric", "min_max_datetime", "min_max_date", "min_max_time", "min_max_string", "min_max_numeric"
]

QueryHookCallable: TypeAlias = "Callable[..., Any]"
InputType = Literal["create", "update_by_pk", "update_by_filter"]
FunctionInfo: TypeAlias = "FilterFunctionInfo | OutputFunctionInfo"

DataclassGraphQLDTO: TypeAlias = "MappedDataclassGraphQLDTO[T] | UnmappedDataclassGraphQLDTO[T]"
GraphQLDTO: TypeAlias = "DataclassGraphQLDTO[T] | MappedPydanticGraphQLDTO[T]"
MappedGraphQLDTO: TypeAlias = "MappedDataclassGraphQLDTO[T] | MappedPydanticGraphQLDTO[T]"
UnmappedGraphQLDTO: TypeAlias = "UnmappedDataclassGraphQLDTO[T]"
AnyMappedDTO: TypeAlias = "MappedDataclassGraphQLDTO[Any] | MappedPydanticGraphQLDTO[Any]"
