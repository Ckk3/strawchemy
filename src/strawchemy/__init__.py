"""Custom DTO implementation."""

from __future__ import annotations

from .mapper import Strawchemy
from .sqlalchemy.hook import FilterOrderHook, LoadColumnsHook, QueryHookProtocol
from .strawberry import ModelInstance
from .strawberry.repository import StrawchemyAsyncRepository, StrawchemySyncRepository

__all__ = (
    "FilterOrderHook",
    "LoadColumnsHook",
    "ModelInstance",
    "QueryHookProtocol",
    "Strawchemy",
    "StrawchemyAsyncRepository",
    "StrawchemySyncRepository",
)
