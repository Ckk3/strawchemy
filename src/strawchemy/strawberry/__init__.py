from __future__ import annotations

from strawchemy.types import DefaultOffsetPagination

from ._field import StrawchemyCreateUpdateMutationField, StrawchemyDeleteMutationField, StrawchemyField
from ._instance import ModelInstance
from ._utils import default_session_getter

__all__ = (
    "DefaultOffsetPagination",
    "ModelInstance",
    "StrawchemyCreateUpdateMutationField",
    "StrawchemyDeleteMutationField",
    "StrawchemyField",
    "default_session_getter",
)
