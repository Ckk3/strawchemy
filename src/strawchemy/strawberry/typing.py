from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal, TypeAlias

if TYPE_CHECKING:
    from collections.abc import Callable

    from sqlalchemy import Select
    from strawberry import Info
    from strawberry.experimental.pydantic.conversion_types import PydanticModel, StrawberryTypeFromPydantic
    from strawberry.types.base import WithStrawberryObjectDefinition
    from strawchemy.graphql.dto import StrawchemyDTOAttributes
    from strawchemy.sqlalchemy.typing import AnyAsyncSession, AnySyncSession

__all__ = (
    "AnySessionGetter",
    "AsyncSessionGetter",
    "FilterStatementCallable",
    "StrawchemyTypeFromPydantic",
    "StrawchemyTypeWithStrawberryObjectDefinition",
    "SyncSessionGetter",
)

GraphQLType = Literal["input", "object", "interface", "enum"]
AsyncSessionGetter: TypeAlias = "Callable[[Info[Any, Any]], AnyAsyncSession]"
SyncSessionGetter: TypeAlias = "Callable[[Info[Any, Any]], AnySyncSession]"
AnySessionGetter: TypeAlias = "AsyncSessionGetter | SyncSessionGetter"
FilterStatementCallable: TypeAlias = "Callable[[Info[Any, Any]], Select[tuple[Any]]]"


if TYPE_CHECKING:

    class StrawchemyTypeWithStrawberryObjectDefinition(StrawchemyDTOAttributes, WithStrawberryObjectDefinition): ...

    class StrawchemyTypeFromPydantic(StrawchemyDTOAttributes, StrawberryTypeFromPydantic[PydanticModel]): ...
