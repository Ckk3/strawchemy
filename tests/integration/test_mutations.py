from __future__ import annotations

from uuid import uuid4

import pytest
from strawchemy import StrawchemyAsyncRepository, StrawchemySyncRepository

import strawberry
from syrupy.assertion import SnapshotAssertion
from tests.typing import AnyQueryExecutor
from tests.utils import maybe_async

from .fixtures import QueryTracker
from .types import ColorCreateInput, ColorType, strawchemy

pytestmark = [pytest.mark.integration]


@strawberry.type
class AsyncMutation:
    create_color: ColorType = strawchemy.create_mutation(ColorCreateInput, repository_type=StrawchemyAsyncRepository)
    create_colors: list[ColorType] = strawchemy.create_mutation(
        ColorCreateInput, repository_type=StrawchemyAsyncRepository
    )


@strawberry.type
class SyncMutation:
    create_color: ColorType = strawchemy.create_mutation(ColorCreateInput, repository_type=StrawchemySyncRepository)
    create_colors: list[ColorType] = strawchemy.create_mutation(
        ColorCreateInput, repository_type=StrawchemySyncRepository
    )


@strawberry.type
class SyncQuery:
    @strawberry.field
    def hello(self) -> str:
        return "world"


@pytest.fixture
def sync_query() -> type[SyncQuery]:
    return SyncQuery


@pytest.fixture
def async_query() -> type[SyncQuery]:
    return SyncQuery


@pytest.fixture
def sync_mutation() -> type[SyncMutation]:
    return SyncMutation


@pytest.fixture
def async_mutation() -> type[AsyncMutation]:
    return AsyncMutation


@pytest.mark.snapshot
async def test_create_mutation(
    any_query: AnyQueryExecutor, query_tracker: QueryTracker, sql_snapshot: SnapshotAssertion
) -> None:
    color_id = uuid4()
    result = await maybe_async(
        any_query(f'mutation {{ createColor(data: {{ id: "{color_id}", name: "new color" }}) {{ id name }} }}')
    )
    assert not result.errors
    assert result.data
    assert result.data["createColor"] == {"id": str(color_id), "name": "new color"}

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.snapshot
async def test_create_many_mutation(
    any_query: AnyQueryExecutor, query_tracker: QueryTracker, sql_snapshot: SnapshotAssertion
) -> None:
    color_id_1, color_id_2 = uuid4(), uuid4()
    result = await maybe_async(
        any_query(
            f"""
                mutation {{
                    createColors(
                        data: [
                            {{ id: "{color_id_1}", name: "new color 1" }}
                            {{ id: "{color_id_2}", name: "new color 2" }}
                        ]
                    ) {{
                        id
                        name
                    }}
                }}
            """
        )
    )
    assert not result.errors
    assert result.data
    assert result.data["createColors"] == [
        {"id": str(color_id_1), "name": "new color 1"},
        {"id": str(color_id_2), "name": "new color 2"},
    ]

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot
