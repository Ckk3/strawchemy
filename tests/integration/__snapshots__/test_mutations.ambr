# serializer version: 1
# name: test_create_many_mutation[session-tracked-async-asyncpg_engine]
  '''
  INSERT INTO color (name, id, created_at, updated_at)
  VALUES ($1::VARCHAR, $2::UUID, $3::TIMESTAMP WITH TIME ZONE, $4::TIMESTAMP WITH TIME ZONE)
  '''
# ---
# name: test_create_many_mutation[session-tracked-async-psycopg_async_engine]
  '''
  INSERT INTO color (name, id, created_at, updated_at)
  VALUES (%(name)s::VARCHAR, %(id)s::UUID, %(created_at)s::TIMESTAMP WITH TIME ZONE, %(updated_at)s::TIMESTAMP WITH TIME ZONE)
  '''
# ---
# name: test_create_many_mutation[session-tracked-sync-psycopg_engine]
  '''
  INSERT INTO color (name, id, created_at, updated_at)
  VALUES (%(name)s::VARCHAR, %(id)s::UUID, %(created_at)s::TIMESTAMP WITH TIME ZONE, %(updated_at)s::TIMESTAMP WITH TIME ZONE)
  '''
# ---
# name: test_create_mutation[session-tracked-async-asyncpg_engine]
  '''
  INSERT INTO color (name, id, created_at, updated_at)
  VALUES ($1::VARCHAR, $2::UUID, $3::TIMESTAMP WITH TIME ZONE, $4::TIMESTAMP WITH TIME ZONE)
  '''
# ---
# name: test_create_mutation[session-tracked-async-psycopg_async_engine]
  '''
  INSERT INTO color (name, id, created_at, updated_at)
  VALUES (%(name)s::VARCHAR, %(id)s::UUID, %(created_at)s::TIMESTAMP WITH TIME ZONE, %(updated_at)s::TIMESTAMP WITH TIME ZONE)
  '''
# ---
# name: test_create_mutation[session-tracked-sync-psycopg_engine]
  '''
  INSERT INTO color (name, id, created_at, updated_at)
  VALUES (%(name)s::VARCHAR, %(id)s::UUID, %(created_at)s::TIMESTAMP WITH TIME ZONE, %(updated_at)s::TIMESTAMP WITH TIME ZONE)
  '''
# ---
